### DISCUSSION TITLE
Allow generic `AppContext`

### DISCUSSION THREAD

**Discussion by webberwang** [vercel/next.js#9935] [Ideas]
# Feature request

## Is your feature request related to a problem? Please describe.

The suggested method is to use `declare` to modify `AppContext`, but it would be better to have a generic `AppContext<T>` to allow us to attach additional properties, such as Apollo Client etc.

## Describe the solution you'd like

<img width="779" alt="codelab-ui-nestjs____Code_UIB_codelab-ui-nestjs__-_____node_modules_next_dist_pages__app_d_ts" src="https://user-images.githubusercontent.com/26264925/71758533-adba7f80-2e55-11ea-9cad-753ab5440826.png">

Allow `AppContext<T>`

## Describe alternatives you've considered

Use `declare` to override interface

## Additional context

Discussed in this thread https://spectrum.chat/next-js/general/how-would-you-type-this~662bfdb5-b174-4503-94a6-e6ad83bca204

**Comment by thief-sty** [comment]
This wouldn't be really useful since `ctx` is only something that is passed to `getInitialProps`, not something that is shared by successive renders. Keeping the same context between renders would make it truly useful, by allowing passage of information through renders so that no dirty hacks that pass information through the DOM or queries have to be done.

**Comment by webberwang** [comment]
> This wouldn't be really useful since `ctx` is only something that is passed to `getInitialProps`, not something that is shared by successive renders. Keeping the same context between renders would make it truly useful, by allowing passage of information through renders so that no dirty hacks that pass information through the DOM or queries have to be done.

I'm attaching Apollo to each render, so the ctx is there, I'm using the declare approach to fix the TSLint errors.

**Comment by KalvinWei** [comment]
@webberwang Hi, mate. I'm new to Next.js. Can different NextJS pages share the same AppContext (from React Context API)?
